v3/docs/gs/params.md

```diff
@@ -199,8 +204,10 @@ The following KVS runtime parameters are part of the stable API.
 | `transactions.enabled` | `false` | Transaction mode (false==disabled, true==enabled) |
 | `mclass.policy` | `auto` | Media class usage (see discussion below for value strings) |
 | `compression.value.algorithm` | `none` | Value compression method (none, lz4) |
-| `compression.value.min_length` | `12` | Value length above which compression is attempted (bytes) |
 
+!!! tip
+    The `compression.value.algorithm` setting in effect for the KVS can be overridden
+    by a flag in individual `hse_kvs_put()` API calls.
```

Comment:
FYI (from hse.h):
```
 * If compression is on by default for the given kvs, then hse_kvs_put() will
 * attempt to compress the value unless the HSE_KVS_PUT_VCOMP_OFF flag is given.
 * Otherwise, the HSE_KVS_PUT_VCOMP_OFF flag is ignored.
 *
 * If compression is off by default for the given kvs, then hse_kvs_put() will
 * not attempt to compress a value unless the HSE_KVS_PUT_VCOMP_ON flag is
 * given. Otherwise, the HSE_KVS_PUT_VCOMP_ON flag is ignored.
```

Reply:
@tristan957 , can you comment as to why we dropped `value` from the compression settings?  Having these settings behind `value` would be result in a more logical schema should we ever support compression of other objects.

Reply:
@beckerg deemed it pointless. I can add it back, but Greg and I couldn't think of anything else that could be compressable in the future. Did you have any ideas?

Reply:
Agree with Alex, but then "compression" shouldn't be a top-level category.  It should probably should be key.compression, value.compression, whatever.compression, ...

Reply:
> The value of value was pointed out by Steve.

Where is this pointed out?

> Agree with Alex, but then "compression" shouldn't be a top-level category. It should probably should be key.compression, value.compression, whatever.compression, ...

I like this idea better. Alex, if you like it, I will take a look today

Reply:
Also, we can't support more than one algorithm because we currently cannot detect which algo to use to decompress.  So anything more than an on/off switch for value compression is premature....

Reply:
Re my "value of value" comment, the following observation was made by Steve in a different conversation.  I was just giving proper credit.
> Having these settings behind value would be result in a more logical schema should we ever support compression of other objects

Reply:
RE value compression: even if we cannot support more than one algorithm today, the benefit of identifying that one is if we add support for others in the future.  The alternative is a now-and-forever default algorithm, in which case just on/off works for now.

RE remove "value": my point was that there is a benefit in clearly separating value from key compression in case we ever implement specialized key compression in the future.  Coumnar stores commonly implement specialized column compression (e.g., delta encoding, incremental encoding, etc.), and our keys can be thought of as a column in a table and theoretically have different compression encodings. -- though I'll be the first to admit it's highly unlikely.

So I'm not suggesting we add anything for key compression now, just make sure there's room for it in the future without changing the someday-to-be-stable API.

Reply:
Agreed.  The point I was trying to make is that we have a huge amount of code to support choosing the algorithm, yet we cannot simply add a new algorithm to HSE.  This is because we do not currently keep track of the algo with which a value was compressed, so simply adding an algo would prevent you from being able to decode existing values.
Adding a new value algo will be quite a bit of work, and have some interesting semantics (e.g., you could have values compressed by four or five different algos all in the same kvs!)  Not hard work mind you, and I would prefer we laid the proper groundwork before liftoff.

So what we have now is premature, and likely much of it will have to be reworked should we ever want to support an additional algorithm.  If we don't give a knob to select the algo, a bunch of code that does a whole lot of nothing can be removed.  And in the future if we wish to support multiple algos then we introduce the knob after proper consideration.  I don't see how we're ever painted into a corner?  IMO, what paints us into a corner is the introduction of knobs now that haven't yet been clearly worked out.

Anyway, what we have now is fine, esp if we never change it (other than the hierarchical naming).  I kinda like mongodb's config file, it has top-level stanzas for networking, profiling, replication, storage, logging, process stuff, ...  I haven't really looked at our stuff from a high level to see how organized it is...

Reply:
I would be concerned about the expense of that approach considering I can think of several places we do a series of if statements to decode the vtype, seems like it'd be better to carve out a few bits for the algo selection.  You could actually do that with kmd_vtype and a mask if you wanted, so then wouldn't require any additional storage (although I don't know how much space in the omf you give to kmd_vtype).

Reply:
Right.  Point is, in the future we can change the default algo in a compatible way.